### PR TITLE
Implemented Method to Allow File to be Relinked if the User Moved the File, Added Tests to Test the Method

### LINKED ISSUE #9798: If user moved file, it should simply be relinked - [Find Unlinked Files part]
A user might move files in the file directory without JabRef.

Then, JabRef does not find the file:

Image

But it still exists.

Example:

1. Open `test-support\src\manual-tests\issue-9798\issue-9798.bib in JabRef
2. Start "Lookup" -> "Find unlinked files"\
    Image  
3. Now, it finds "minimal.pdf"\
    Image

Current behavior: Blank right of minimal.pdf - and it would be re-imported

Expected result:

File should be relinked - and following shown in the status bar

> file relinked to entry "minimal"

---

Alternative explanation:

Start:

```bibtex
@misc{test,
  file = {a.pdf}
}
```

File `a.pdf` exists

User moves `a.pdf` into `folder-a`.

When "Find unlinked files" is started, JabRef should

- Recognize that `a.pdf` of `test` is missing
- `folder-a/a.pdf` does not belong to any entry
- Rewrite the link of `test` from `a.pdf` to `folder-a/a.pdf`

If `a.pdf` exists multiple times, no rewrite should be done.

## First implementation

The functionality should be in "Automatically set file links" (Quality -> Automatically set file links). In case it is absolutely sure that the file was moved (i.e., no two files with the same name), they should be moved.

image

Current output:

image

## Follow-up implementation (NOT IN SCOPE OF THIS ISSUE)

(The following text is only a reminder of the development team to open a new issue)

The "Search for unlinked files" (Lookup -> Find unlinked files) dialog should get two columns as search result. Currently 1.

- column 1: file x@folder y (as it is currently; with checkmarks)
- column 2: candidate entry (citation key. If clicked, main table jumps to that entry)

In case no candidate entry exists, "<new>" is put in. In case two (or more) candidate entries exist, the cell turns into a dropdown. The jump link is next to the drop down.

image

---

**Update 2024-01-09**: Testing effort **high** is serious. Crafting test cases for different scenarios is necessary. See especially the comment at https://github.com/JabRef/jabref/issues/9798#issuecomment-1761330493.

**Update 2024-03-05**: Testing effort **high** is really serious. A single test case is not enough. The PR https://github.com/JabRef/jabref/pull/10526 showed an initial test case. -- One could think of using https://github.com/google/jimfs for more advanced tests. However `@TempDir` with some files should be enough.

### PR DESCRIPTION
**Pull request by u7754676**
This fix was co-written by [Le-Chat-BottE](https://github.com/Le-Chat-BottE) and I. We have implemented the "Automatically set file links" feature. Specifically, this was the implementation described under the heading "First Implementation" in the original issue #9798. Our implementation comes in the form of the addition of the new method **relinkFileIfMoved** in the **AutoSetFileLinks.java** file.

In addition, as requested in the issue itself, we have also created 6 new tests to test this new method, located in the **AutoSetFileLinksUtilTest.java** file. The 6 tests are as follows:

- relinkMoveFileFromRootFolderToSubfolder
- relinkMoveFileFromSubfolderToRootFolder
- relinkMoveFileFromSubfolderToSubfolder
- noRelinkCopyFileFromRootFolderToSubfolder
- noRelinkCopyFileFromSubfolderToRootFolder
- noRelinkCopyFileFromSubfolderToSubfolder

We think our code is ready for review and we welcome any feedback or suggestions. Thank you!

Fixes #9798.

### Mandatory checks

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java [hunk 1]
```diff
@@ -68,6 +72,20 @@ private AutoSetFileLinksUtil(List<Path> directories, ExternalApplicationsPrefere
         this.filePreferences = filePreferences;
     }
 
+    // If file that is already linked to an existing entry is moved, the file should be relinked to the entry
+    private void relinkFileIfMoved(List<LinkedFile> newLinkedFiles, BibEntry entry) {
+        // Get the list of files linked to a specific entry
+        List<LinkedFile> allLinkedFiles = entry.getField(StandardField.FILE).map(FileFieldParser::parse).orElse(Collections.emptyList());
```
Comment (koppor) [inline src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java:78]: Use entry.getFiles()

**Review by koppor — CHANGES_REQUESTED** [review]
Logik is with flaws, if mutiple files are linked to an entry (e.g., coverage). It does not work.

Please improve.

**Comment by subhramit** [comment]
Hey @u7754676 
The title of the PR should not contain issue numbers, as it becomes hard to follow. It should only tell what the PR is doing.

**Comment by koppor** [comment]
Closing this issue due to inactivity :zzz:
